### PR TITLE
Refactor & finishing first iteration of Character-Data microservice

### PR DESCRIPTION
**Pull request by ArtiKitten**
#### What changes does this PR contain?
First iteration of Character-Data refactored from template changes
#### Which JIRA ticket ID's are associated with this PR?
UBI-337
#### Does this PR change the outward facing API?
Yes
#### Additional comments and information
Some tests were added since the first iteration. There is still some tests to add when it'll talk to the User Microservice to validate that a user exists. It still uses fake DB and int id for the data. We'll need to talk about what kind of character name we allow to edit the validation regex to fit our conditions.
@ubivius/microservices

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,14 @@
-# character-data-microservice
-Character data microservice for our online game framework. 
+# microservice-character-data
+Character Data microservice
+
+Beginning of code was based off of Nic Jackson's building microservices in Go : https://www.youtube.com/playlist?list=PLmD8u-IFdreyh6EUfevBcbiuCKzFk0EW_
+
+## TODO
```
Comment (ArcticWalrus) [inline README.md:6]: remove todo

File: chart/templates/NOTES.txt [hunk 2]
```diff
@@ -0,0 +1,25 @@
+Thank you for installing {{ .Chart.Name }}.
```
Comment (ArcticWalrus) [inline chart/templates/NOTES.txt:1]: Do a search for all mentions of microservice-template and replace with the correct name

File: data/characters.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+import (
+	"fmt"
+	"time"
+)
+
+// Character specific errors
+var ErrorCharacterNotFound = fmt.Errorf("Character not found")
+
+// Character defines the structure for an API character.
+// Formatting done with json tags to the right. "-" : don't include when encoding to json
```
Comment (ArcticWalrus) [inline data/characters.go:12]: Remove comment about json tags

File: data/characters.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+
+// DELETING A CHARACTER
+func DeleteCharacter(id int) error {
+	index := findIndexByCharacterID(id)
+	if index == -1 {
+		return ErrorCharacterNotFound
+	}
+
+	// This should not work, probably needs ':' after index+1. To test
+	characterList = append(characterList[:index], characterList[index+1])
```
Comment (ArcticWalrus) [inline data/characters.go:68]: Fix delete as in microservice-template

File: data/json.go [hunk 5]
```diff
@@ -0,0 +1,18 @@
+package data
+
+import (
+	"encoding/json"
+	"io"
+)
+
+// FromProductJSON deserializes the interface from JSON string
+func FromJSON(i interface{}, reader io.Reader) error {
```
Comment (ArcticWalrus) [inline data/json.go:9]: Remove From and To JSON

File: handlers/middleware.go [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/Ubivius/microservice-character-data/data"
+)
+
+// Errors should be templated in the future.
```
Comment (ArcticWalrus) [inline handlers/middleware.go:11]: Remove this comment from non microservice-template repo

File: main.go [hunk 7]
```diff
@@ -1,48 +1,72 @@
...
+			logger.Fatal(err)
+		}
+	}()
 
-	// Handlers
-	helloHandler := handlers.NewHello(l)
-	achievementHandlers := handlers.NewAchievement(l)
+	// Handle shutdown signals from operating system
+	signalChannel := make(chan os.Signal)
+	signal.Notify(signalChannel, os.Interrupt)
```
Comment (ArcticWalrus) [inline main.go:63]: These linting errors are fixed in microservice-template

File: pkg/data/characters.go [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+	"fmt"
+	"time"
+)
+
+// Character specific errors
+var ErrorCharacterNotFound = fmt.Errorf("Character not found")
+var ErrorCharacterNameAlreadyExist = fmt.Errorf("Character name is already used")
+
+// Character defines the structure for an API character.
+// Formatting done with json tags to the right. "-" : don't include when encoding to json
```
Comment (ArcticWalrus) [inline pkg/data/characters.go:13]: Same thing for comment here

File: pkg/data/characters.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+// DELETING A CHARACTER
+func DeleteCharacter(id int) error {
+	index := FindIndexByCharacterID(id)
+	if index == -1 {
+		return ErrorCharacterNotFound
+	}
+
+	// This should not work, probably needs ':' after index+1. To test
+	characterList = append(characterList[:index], characterList[index+1])
```
Comment (ArcticWalrus) [inline pkg/data/characters.go:75]: Fix delete

File: pkg/data/json.go [hunk 10]
```diff
@@ -0,0 +1,18 @@
+package data
+
+import (
+	"encoding/json"
+	"io"
+)
+
+// FromProductJSON deserializes the interface from JSON string
```
Comment (ArcticWalrus) [inline pkg/data/json.go:8]: Same file twice. Why do we have json in pkg and data

File: pkg/server/server.go [hunk 11]
```diff
@@ -0,0 +1 @@
+package server
```
Comment (ArcticWalrus) [inline pkg/server/server.go:1]: We can remove this package until the refactor is actually done in microservice-template

File: scripts/elasticsearch.sh [hunk 12]
```diff
@@ -0,0 +1,30 @@
+#!/bin/bash
```
Comment (ArcticWalrus) [inline scripts/elasticsearch.sh:1]: This script is useless here. We can remove it

**Review by ArcticWalrus — CHANGES_REQUESTED** [review]

File: pkg/data/json.go (hunk 10 above)
Comment (ArtiKitten) [reply pkg/data/json.go:8]: Pour une raison obscure, après avoir modifier le format pour avoir pkd/data, pkg/handlers, etc. Je n'ai pas delete les dossiers originaux.

File: chart/templates/NOTES.txt (hunk 2 above)
Comment (ArtiKitten) [reply chart/templates/NOTES.txt:1]: Selon @MaximeStOnge, il faudrait une autre tâche pour setup les charts pour le déploiement

File: main.go (hunk 7 above)
Comment (ArtiKitten) [reply main.go]: What should I do about this?

File: pkg/router/router.go [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+	"log"
+	"net/http"
+
+	"github.com/Ubivius/microservice-character-data/pkg/handlers"
+	"github.com/gorilla/mux"
+)
+
+// Mux route handling with gorilla/mux
+func New(charactersHandler *handlers.CharactersHandler, logger *log.Logger) *mux.Router {
+	// Mux route handling with gorilla/mux
```
Comment (MaximeStOnge) [inline pkg/router/router.go:13]: Duplicate comment

File: pkg/data/characters.go [hunk 14]
```diff
@@ -0,0 +1,133 @@
+package data
+
+import (
+	"fmt"
+	"time"
+)
+
+// Character specific errors
+var ErrorCharacterNotFound = fmt.Errorf("Character not found")
+var ErrorCharacterNameAlreadyExist = fmt.Errorf("Character name is already used")
```
Comment (MaximeStOnge) [inline pkg/data/characters.go:10]: Est-ce qu'on veut vraiment que les noms de personnage soit unique?

File: pkg/handlers/handlers_test.go [hunk 15]
```diff
@@ -0,0 +1,183 @@
...
+
+	characterHandler := NewCharactersHandler(NewTestLogger())
+	characterHandler.AddCharacter(response, request)
+
+	if response.Code != http.StatusConflict {
+		t.Errorf("Expected status code %d, but got %d", http.StatusConflict, response.Code)
+	}
+}
+
+func TestUpdateCharacter(t *testing.T) {
```
Comment (MaximeStOnge) [inline pkg/handlers/handlers_test.go:144]: Il faudrait un TestUpdateNonExistantCharacter

**Review by MaximeStOnge — CHANGES_REQUESTED** [review]

File: pkg/data/characters.go (hunk 14 above)
Comment (ArtiKitten) [reply pkg/data/characters.go:10]: À discuter avec les autres. Au pire, on l'enlèvera, ce n'est pas un problème.

**Review by MaximeStOnge — APPROVED** [review]

File: main.go (hunk 7 above)
Comment (ArcticWalrus) [reply main.go]: https://github.com/Ubivius/microservice-template/blob/master/cmd/microservice-template/main.go#L69
and under

File: README.md [hunk 16]
```diff
@@ -1,2 +1,4 @@
-# character-data-microservice
-Character data microservice for our online game framework. 
+# microservice-character-data
+Character Data microservice
+
+Beginning of code was based off of Nic Jackson's building microservices in Go : https://www.youtube.com/playlist?list=PLmD8u-IFdreyh6EUfevBcbiuCKzFk0EW_
```
Comment (ArcticWalrus) [inline README.md:4]: Remove Nic Jackson reference

**Review by ArcticWalrus — CHANGES_REQUESTED** [review]

**Review by ArcticWalrus — APPROVED** [review]

**Pull request merged**